<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A streamlined workflow for long-read DNA methylation analysis with NanoMethViz and Bioconductor • LongMod123</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="A streamlined workflow for long-read DNA methylation analysis with NanoMethViz and Bioconductor">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LongMod123</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/workflow.html">A streamlined workflow for long-read DNA methylation analysis with NanoMethViz and Bioconductor</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/shians/differential_methylation_workshop" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>A streamlined workflow for long-read DNA
methylation analysis with NanoMethViz and Bioconductor</h1>
                        <h4 data-toc-skip class="author">Shian
Su<sup>1</sup>, Lucinda Xiao<sup>1</sup>, James Lancaster<sup>1</sup>,
Tamara Cameron<sup>1</sup>, Kelsey Breslin<sup>1</sup>, Peter F.
Hickey<sup>1</sup>, Marnie E. Blewitt<sup>1</sup>, Quentin
Gouil<sup>1,2</sup>, Matthew E. Ritchie<sup>1</sup>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mritchielab/LongMod123/blob/main/vignettes/workflow.rmd" class="external-link"><code>vignettes/workflow.rmd</code></a></small>
      <div class="hidden name"><code>workflow.rmd</code></div>

    </div>

    
    
<p><sup>1</sup>The Walter and Eliza Hall Institute of Medical Research,
Parkville, Victoria, Australia and Department of Medical Biology, The
University of Melbourne, Parkville, Victoria, Australia.</p>
<p><sup>2</sup>School of Life and Environmental Sciences, University of
Sydney, Sydney, New South Wales, Australia.</p>
<div class="section level2">
<h2 id="abstract">Abstract<a class="anchor" aria-label="anchor" href="#abstract"></a>
</h2>
<p>The development of long-read sequencing technologies has introduced a
powerful new tool to the field of epigenetics by enabling the direct
detection of DNA modifications, such as DNA methylation, at single-base
resolution. This produces novel opportunities for studying the role of
DNA methylation in gene regulation, imprinting, and disease. Analysis of
the data produced by long-read sequencing technologies requires the
development of new software tools to handle the unique characteristics
of the data, such as the modBAM format in which the data is output, and
the length of the sequenced reads. The <code>NanoMethViz</code> package
provides a suite of tools for loading in this methylation data,
visualising regional and genome-wide methylation patterns, and exporting
data for identifying differentially methylated regions (DMRs) using
other Bioconductor software such as <code>bsseq</code>,
<code>DSS</code>, <code>dmrseq</code> and <code>edgeR</code>. In this
workflow article, we demonstrate how to use the <code>NanoMethViz</code>
package and other Bioconductor software to perform a complete analysis
of long-read 5mC DNA methylation data starting from the raw data, to
exploratory analysis, DMR identification, and visualisation of the
results.</p>
</div>
<div class="section level2">
<h2 id="keywords">Keywords<a class="anchor" aria-label="anchor" href="#keywords"></a>
</h2>
<p>DNA methylation, Differential methylation, Long-read sequencing,
Epigenetics, Bioconductor</p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>DNA methylation is a key epigenetic regulator of gene expression in
mammals, involving the addition of a methyl group to cytosine or adenine
nucleotides. In mammals, the predominant form of DNA methylation is
5-methylcytosine (5mC) in CpG dinucleotides. CpG methylation is involved
in many important epigenetic processes, including differential gene
expression, X-chromosome inactivation, genomic imprinting, and
maintenance of genome stability <span class="citation">(1)</span>.
Changes in CpG methylation are implicated in a variety of different
disease processes, including tumorigenesis and imprinting disorders such
as Prader-Willi Syndrome and Angelman Syndrome <span class="citation">(2)</span>.</p>
<p>Over the past couple of decades, multiple methods have been developed
to measure DNA methylation, including bisulfite sequencing <span class="citation">(3)</span>, methylation arrays <span class="citation">(4)</span>, and more recently, third-generation
sequencing technologies such as those offered by Oxford Nanopore
Technologies (ONT) and Pacific Biosciences (PacBio). While bisulfite
sequencing has long been the gold standard for DNA methylation analysis,
it suffers from several limitations; these include DNA degradation from
the process of bisulfite conversion, incomplete conversion of
unmethylated cytosines, amplification bias from PCR, and short read
lengths that pose difficulties for assembly and haplotyping <span class="citation">(5)</span>. By contrast, ONT’s nanopore sequencing and
PacBio’s single-molecule real-time (SMRT) sequencing can directly
sequence DNA molecules with all their modifications intact, and use
basecalling algorithms to predict DNA modifications from the raw
sequencing signal. Basecalling algorithms have the potential for bias,
depending on the algorithm and training data used for the model;
however, improvements are continually being made to generate more
accurate models, and additionally, to detect other DNA modifications in
other sequence contexts <span class="citation">(6)</span>.</p>
<p>In this workflow, we will demonstrate how to use the
<code>NanoMethViz</code> package <span class="citation">(7)</span> and
other Bioconductor software to visualise and analyse methylation data
generated by ONT and PacBio long-read sequencing. The sequencing data is
currently output in the modBAM file format, which contains the
basecalled reads alongside tags for modification information. The
<code>NanoMethViz</code> package provides a suite of tools for loading
in this methylation data, visualising regional and genome-wide
methylation patterns, and exporting data for identifying differentially
methylated regions (DMRs) using other Bioconductor software such as
<code>bsseq</code> <span class="citation">(8)</span>, <code>DSS</code>
<span class="citation">(9)</span>, <code>dmrseq</code> <span class="citation">(10)</span> and <code>edgeR</code> <span class="citation">(11)</span> (Figure @ref(fig:fig1)B).</p>
</div>
<div class="section level2">
<h2 id="experimental-setup">Experimental Setup<a class="anchor" aria-label="anchor" href="#experimental-setup"></a>
</h2>
<p>In this workflow, we will analyse ultra-long-read Nanopore sequencing
data from the neural stem cells of triplicate E14.5 female mice <span class="citation">(12)</span>. The mice were F1 crosses between an Xist
knockout (X<sup>ΔA</sup>X) female <span class="citation">(13)</span> and
a Castaneus strain (CAST) male. The resulting offspring have
sufficiently different parental chromosomes to allow effective
genome-wide haplotyping to distinguish the parent-of-origin for
long-reads, allowing for the identification of genes that are imprinted
or strain-specific. The Xist KO of the maternal genome also guarantees
that X-inactivation takes place on the paternal X chromosome, allowing
for the study of X-inactivation. The DNA was sequenced on R9.4.1
PromethION flow cells, basecalled with Dorado, aligned to the mm10 mouse
reference genome, and phased with WhatsHap <span class="citation">(14)</span>, to produce the phased modBAM (Figure
@ref(fig:fig1)A). For illustrative purposes, we use a we use a subset of
the data focusing on chromosome 7, which contains several known
imprinted genes.</p>
<div class="figure">
<img src="Figure1.png" alt="**Overview of the workflow**. (A) Experimental setup for the data used in this workflow, from Gocuk and Lancaster et al., 2024. (B) Flowchart of the workflow for long-read methylation data visualisation and analysis using `NanoMethViz` and other Bioconductor packages including `limma`, `dmrseq`, `bsseq`, and `edgeR`. (C) Summary of data structures for the storage of methylation data used in different parts of the workflow. Xi: inactive X. Xa: active X. CAST: Castaneus. 5mC: 5-methylcytosine. MDS: multidimensional scaling. DMR: differentially methylated region. Figure created with [BioRender.com](BioRender.com)." width="600"><p class="caption">
<strong>Overview of the workflow</strong>. (A) Experimental setup for
the data used in this workflow, from Gocuk and Lancaster et al., 2024.
(B) Flowchart of the workflow for long-read methylation data
visualisation and analysis using <code>NanoMethViz</code> and other
Bioconductor packages including <code>limma</code>, <code>dmrseq</code>,
<code>bsseq</code>, and <code>edgeR</code>. (C) Summary of data
structures for the storage of methylation data used in different parts
of the workflow. Xi: inactive X. Xa: active X. CAST: Castaneus. 5mC:
5-methylcytosine. MDS: multidimensional scaling. DMR: differentially
methylated region. Figure created with <a href="BioRender.com">BioRender.com</a>.
</p>
</div>
</div>
<div class="section level2">
<h2 id="loading-in-data">Loading in data<a class="anchor" aria-label="anchor" href="#loading-in-data"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BiocFileCache</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">dmrseq</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/shians/NanoMethViz" class="external-link">NanoMethViz</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span></code></pre></div>
<p>We begin by loading the packages and downloading the input data which
contains the modBAM files required for this workflow.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="st">"input"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">cache_path</span> <span class="op">&lt;-</span> <span class="fu">tools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/tools/userdir.html" class="external-link">R_user_dir</a></span><span class="op">(</span><span class="st">"LongMod123"</span>, which <span class="op">=</span> <span class="st">"cache"</span><span class="op">)</span></span>
<span>    <span class="va">bfc</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">BiocFileCache</a></span><span class="op">(</span><span class="va">cache_path</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    <span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://zenodo.org/records/12747551/files/input.tar.gz?download=1"</span></span>
<span>    </span>
<span>    <span class="va">qry</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcquery</a></span><span class="op">(</span><span class="va">bfc</span>, <span class="va">url</span>, exact <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">qry</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">input_path</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span><span class="va">bfc</span>, <span class="st">"input"</span>, fpath <span class="op">=</span> <span class="va">url</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="kw">if</span> <span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcneedsupdate</a></span><span class="op">(</span><span class="va">bfc</span>, <span class="va">qry</span><span class="op">$</span><span class="va">rid</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcdownload</a></span><span class="op">(</span><span class="va">bfc</span>, <span class="va">qry</span><span class="op">$</span><span class="va">rid</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>        <span class="va">input_path</span> <span class="op">&lt;-</span> <span class="va">bfc</span><span class="op">[[</span><span class="va">qry</span><span class="op">$</span><span class="va">rid</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span><span class="va">input_path</span>, exdir <span class="op">=</span> <span class="st">"."</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>As mentioned, modBAM files are the currently preferred output of
Dorado, the primary methylation caller offered by ONT. modBAM files are
BAM files containing reads with additional tags on each read storing
information about methylation calls (the MM tag) alongside their
methylation probabilities (the ML tag). To use this data in
<code>NanoMethViz</code>, we must construct an object of the
<code>ModBamResults</code> class.</p>
<p>The <code>ModBamResults</code> class is a container for storing
methylation data from multiple samples, along with sample annotation and
exon annotation. To create this object, we need to provide paths to the
modBAM files, a sample table, and optionally exon annotation.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># List the modBAM files by searching for files with the .bam extension in the input directory</span></span>
<span><span class="va">bam_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="st">"input"</span>, pattern <span class="op">=</span> <span class="st">"*bam$"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">bam_files</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "input/nsc_1_mat.bam" "input/nsc_1_pat.bam" "input/nsc_2_mat.bam"</span></span>
<span><span class="co">## [4] "input/nsc_2_pat.bam" "input/nsc_3_mat.bam" "input/nsc_3_pat.bam"</span></span></code></pre>
<p>The samples need to be annotated with information for
<code>NanoMethViz</code> to use in aggregating data within experimental
groups. The sample table must contain <code>sample</code> and
<code>group</code> columns with optional additional columns for further
annotation. The rows of this table must match the order of the input
files. The <code>group</code> column is generally the default grouping
column used in the <code>NanoMethViz</code> package. In this example, we
have grouped our data by haplotype.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">"input/sample_anno.tsv"</span>, show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">samples</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">##   sample    group   tissue haplotype</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> nsc_1_mat nsc_mat nsc    mat      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> nsc_1_pat nsc_pat nsc    pat      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> nsc_2_mat nsc_mat nsc    mat      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> nsc_2_pat nsc_pat nsc    pat      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> nsc_3_mat nsc_mat nsc    mat      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> nsc_3_pat nsc_pat nsc    pat</span></span></code></pre>
<p>Optionally we can add exon annotation to the object to generate
tracks for gene annotation when plotting genomic regions/genes. The exon
annotation must be a <code>data.frame</code> with columns
<code>gene_id</code>, <code>chr</code>, <code>strand</code>,
<code>start</code>, <code>end</code>, <code>transcript_id</code>, and
<code>symbol</code>. A number of helper functions are provided to
retrieve exon annotations in the correct format for human (hg19, hg38)
and mouse (mm10, GRCm39). In this example, we will use the
<code><a href="https://rdrr.io/pkg/NanoMethViz/man/get_exons.html" class="external-link">get_exons_mm10()</a></code> function to retrieve exon annotations for
the mouse genome (mm10) and filter for chromosome 7.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exon_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/get_exons.html" class="external-link">get_exons_mm10</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">chr</span> <span class="op">==</span> <span class="st">"chr7"</span><span class="op">)</span></span>
<span><span class="va">exon_anno</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 60,146 × 7</span></span></span>
<span><span class="co">##    gene_id   chr   strand    start      end transcript_id symbol </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> 100009609 chr7  -      84<span style="text-decoration: underline;">935</span>565 84<span style="text-decoration: underline;">941</span>088         <span style="text-decoration: underline;">60</span>883 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> 100009609 chr7  -      84<span style="text-decoration: underline;">943</span>141 84<span style="text-decoration: underline;">943</span>264         <span style="text-decoration: underline;">60</span>883 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> 100009609 chr7  -      84<span style="text-decoration: underline;">943</span>504 84<span style="text-decoration: underline;">943</span>722         <span style="text-decoration: underline;">60</span>883 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> 100009609 chr7  -      84<span style="text-decoration: underline;">946</span>200 84<span style="text-decoration: underline;">947</span>000         <span style="text-decoration: underline;">60</span>883 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> 100009609 chr7  -      84<span style="text-decoration: underline;">947</span>372 84<span style="text-decoration: underline;">947</span>651         <span style="text-decoration: underline;">60</span>883 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> 100009609 chr7  -      84<span style="text-decoration: underline;">963</span>816 84<span style="text-decoration: underline;">964</span>115         <span style="text-decoration: underline;">60</span>883 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> 100009609 chr7  -      84<span style="text-decoration: underline;">935</span>565 84<span style="text-decoration: underline;">941</span>088         <span style="text-decoration: underline;">60</span>884 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> 100009609 chr7  -      84<span style="text-decoration: underline;">943</span>141 84<span style="text-decoration: underline;">943</span>264         <span style="text-decoration: underline;">60</span>884 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> 100009609 chr7  -      84<span style="text-decoration: underline;">943</span>504 84<span style="text-decoration: underline;">943</span>722         <span style="text-decoration: underline;">60</span>884 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> 100009609 chr7  -      84<span style="text-decoration: underline;">946</span>200 84<span style="text-decoration: underline;">947</span>000         <span style="text-decoration: underline;">60</span>884 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 60,136 more rows</span></span></span></code></pre>
<p>Once we have created the <code>ModBamResult</code> object, we can use
the <code><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene()</a></code> function to visualise the methylation
patterns of a gene of interest. In this example, we will plot the
methylation patterns of the <em>Peg3</em> (paternally-expressed gene 3)
gene on chromosome 7, which is known to be imprinted in mice. The plot
contains a smoothed trendline of the group-aggregated methylation
probabilities across the gene, along with a heatmap showing the
methylation probabilities of individual CpG sites in each read, and a
track showing the isoforms of the gene if exon annotation is provided
(Figure @ref(fig:fig2-gene-peg3)).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mbr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/ModBamResult-class.html" class="external-link">ModBamResult</a></span><span class="op">(</span></span>
<span>    methy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/ModBamFiles.html" class="external-link">ModBamFiles</a></span><span class="op">(</span></span>
<span>        paths <span class="op">=</span> <span class="va">bam_files</span>,</span>
<span>        samples <span class="op">=</span> <span class="va">samples</span><span class="op">$</span><span class="va">sample</span></span>
<span>    <span class="op">)</span>,</span>
<span>    samples <span class="op">=</span> <span class="va">samples</span>,</span>
<span>    exons <span class="op">=</span> <span class="va">exon_anno</span>,</span>
<span>    mod_code <span class="op">=</span> <span class="st">"m"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene</a></span><span class="op">(</span><span class="va">mbr</span>, <span class="st">"Peg3"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="workflow_files/figure-html/fig2-gene-peg3-1.png" alt="**Gene-level plot of Peg3**. This plot shows the methylation information along the *Peg3* gene on chromosome 7. The top track shows a rolling-smoothed average of methylation probability across the region, aggregated across samples within each experimental group. The middle track shows a heatmap of the methylation probabilities within individual reads, separated by experimental group. The bottom track shows isoform annotations for *Peg3*, with rectangles representing exons, lines representing introns and arrows representing the direction of transcription." width="600"><p class="caption">
<strong>Gene-level plot of Peg3</strong>. This plot shows the
methylation information along the <em>Peg3</em> gene on chromosome 7.
The top track shows a rolling-smoothed average of methylation
probability across the region, aggregated across samples within each
experimental group. The middle track shows a heatmap of the methylation
probabilities within individual reads, separated by experimental group.
The bottom track shows isoform annotations for <em>Peg3</em>, with
rectangles representing exons, lines representing introns and arrows
representing the direction of transcription.
</p>
</div>
<p>CpG dinucleotides are often enriched around gene promoters, where
methylation is generally correlated with silencing and hypomethylation
with activation. In the plot, we see that there is a strong
demethylation pattern on the paternal allele near the transcription
start site (TSS) of the gene, allowing the gene to be expressed. The
maternal allele is hypermethylated near the TSS, silencing the gene.
This is a typical methylation pattern for an imprinted gene, and
consistent with paternal expression of <em>Peg3</em>.</p>
</div>
<div class="section level2">
<h2 id="genome-wide-analysis">Genome-wide analysis<a class="anchor" aria-label="anchor" href="#genome-wide-analysis"></a>
</h2>
<p>While it is possible to use modBAM files to analyse specific regions
of interest, genome-wide analysis requires the data to be in a
tabix-indexed TSV file as it is easier to parse at scale (Figure C). We
can convert the modBAM files to a tabix-indexed TSV file using the
<code><a href="https://rdrr.io/pkg/NanoMethViz/man/modbam_to_tabix.html" class="external-link">modbam_to_tabix()</a></code> function. This function will create a
tabix-indexed TSV file containing the methylation data from the modBAM
files. This file can then be used as input to the
<code>NanoMethViz</code> functions that require sorted genomic data. In
this example, we will convert the <code>ModBamResult</code> object to a
tabix-indexed TSV file and save it to the <code>data</code> directory.
We will use a pre-generated file in the interest of time, but you can
run the code below to generate the file.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"data"</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"data/methy.tsv.bgz"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/modbam_to_tabix.html" class="external-link">modbam_to_tabix</a></span><span class="op">(</span><span class="va">mbr</span>, <span class="st">"data/methy.tsv.bgz"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Once we have the tabix-indexed TSV file, we can create a
<code>NanoMethResult</code> object using the
<code><a href="https://rdrr.io/pkg/NanoMethViz/man/NanoMethResult-class.html" class="external-link">NanoMethResult()</a></code> function. The <code>NanoMethResult</code>
object contains the same information as the <code>ModBamResult</code>
object but with the methylation data stored in a tabix-indexed TSV file.
To create this object we need to provide paths to the tabix-indexed TSV
file, along with the accompanying sample table and exon annotation we
generated before.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nmr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/NanoMethResult-class.html" class="external-link">NanoMethResult</a></span><span class="op">(</span></span>
<span>    methy <span class="op">=</span> <span class="st">"data/methy.tsv.bgz"</span>,</span>
<span>    samples <span class="op">=</span> <span class="va">samples</span>,</span>
<span>    exons <span class="op">=</span> <span class="va">exon_anno</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The <code>NanoMethResult</code> object behaves in the same way as
<code>ModBamResult</code> for any plotting functions. For example we can
use <code><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene()</a></code> to visualise the methylation patterns of
the <em>Peg3</em> gene in exactly the same way but replacing the
<code>ModBamResult</code> object with the <code>NanoMethResult</code>
object. We should see the exact same plot as before.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene</a></span><span class="op">(</span><span class="va">nmr</span>, <span class="st">"Peg3"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="aggregating-feature-methylation-data">Aggregating feature methylation data<a class="anchor" aria-label="anchor" href="#aggregating-feature-methylation-data"></a>
</h2>
<p>It is often informative to aggregate methylation data over a class of
features to identify broad patterns of methylation across the features.
This can help establish overall differences in methylation patterns
between groups in a class of features such as genes, CpG islands, or
enhancers. The <code>NanoMethViz</code> package provides a
<code><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_agg_regions.html" class="external-link">plot_agg_regions()</a></code> function for visualising methylation
data that has been aggregated across a set of genomic regions of
interest, defined by a table of coordinates.</p>
<p>For example, we may want to investigate the methylation patterns of a
set of genes on chromosome 7. We can use the
<code><a href="https://rdrr.io/pkg/NanoMethViz/man/exons_to_genes.html" class="external-link">exons_to_genes()</a></code> helper function to convert exon
annotations already stored in the object to gene annotations and filter
for only chromosome 7 genes (Figure @ref(fig:fig3-gene-agg)). For faster
processing, we subset 200 genes and use the
<code><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_agg_regions.html" class="external-link">plot_agg_regions()</a></code> function to visualise the mean
methylation proportions along the sampled genes. By default, the
methylation proportion is calculated using a threshold of 0.5
modification probability.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/exons_to_genes.html" class="external-link">exons_to_genes</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/exons.html" class="external-link">exons</a></span><span class="op">(</span><span class="va">nmr</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_agg_regions.html" class="external-link">plot_agg_regions</a></span><span class="op">(</span><span class="va">nmr</span>, regions <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_sample</a></span><span class="op">(</span><span class="va">gene_anno</span>, n <span class="op">=</span> <span class="fl">200</span><span class="op">)</span>, group_col <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="workflow_files/figure-html/fig3-gene-agg-1.png" alt="**Aggregated gene-level plot**. The mean methylation proportion is shown across a sample of 200 genes in chromosome 7, aggregated by experimental group. Methylation proportions are individual CpG sites are calculated along relative genomic positions, with fixed flanking regions, then averaged across all genes in each experimental group to produce a smoothed trendline. The x-axis label gives the average width and standard deviation of widths of the features used." width="600"><p class="caption">
<strong>Aggregated gene-level plot</strong>. The mean methylation
proportion is shown across a sample of 200 genes in chromosome 7,
aggregated by experimental group. Methylation proportions are individual
CpG sites are calculated along relative genomic positions, with fixed
flanking regions, then averaged across all genes in each experimental
group to produce a smoothed trendline. The x-axis label gives the
average width and standard deviation of widths of the features used.
</p>
</div>
<p>In this plot we can see that active genes on chromosome 7 tend to be
demethylated near the TSS region and hypermethylated throughout the gene
body. This is consistent with the proposed function of methylation in
gene regulation, where methylation in promoter region of genes is
associated with gene silencing, while transcription of active genes is
associated with recruitment of methylating mechanisms to the gene
body.</p>
<p>Another class of features of interest are CpG islands, which are
regions of the genome that are rich in CpG dinucleotides. CpG islands
are often associated with gene promoters and are predominantly
unmethylated in normal cells <span class="citation">(15)</span>. We can
use the <code><a href="https://rdrr.io/pkg/NanoMethViz/man/get_exons.html" class="external-link">get_cgi_mm10()</a></code> function to retrieve CpG island
annotations for the mouse genome (mm10) and filter for chromosome 7. We
can then use the <code><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_agg_regions.html" class="external-link">plot_agg_regions()</a></code> function to visualise
the mean methylation profile across CpG islands on chromosome 7 (Figure
@ref(fig:fig4-cgi-agg)).</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cgi_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/get_exons.html" class="external-link">get_cgi_mm10</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">chr</span> <span class="op">==</span> <span class="st">"chr7"</span><span class="op">)</span></span>
<span><span class="va">cgi_anno</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1,193 × 14</span></span></span>
<span><span class="co">##      bin chr     start     end gene_id  length cpgNum gcNum perCpg perGc obsExp</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span>   609 chr7  3<span style="text-decoration: underline;">181</span>031 3<span style="text-decoration: underline;">181</span>267 CpG: 22     236     22   147   18.6  62.3   1   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span>   610 chr7  3<span style="text-decoration: underline;">303</span>442 3<span style="text-decoration: underline;">303</span>817 CpG: 28     375     28   249   14.9  66.4   0.69</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span>   610 chr7  3<span style="text-decoration: underline;">366</span>695 3<span style="text-decoration: underline;">366</span>942 CpG: 27     247     27   152   21.9  61.5   1.22</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span>   611 chr7  3<span style="text-decoration: underline;">411</span>983 3<span style="text-decoration: underline;">412</span>269 CpG: 26     286     26   188   18.2  65.7   0.84</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span>   611 chr7  3<span style="text-decoration: underline;">414</span>874 3<span style="text-decoration: underline;">415</span>855 CpG: 87     981     87   518   17.7  52.8   1.28</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span>   611 chr7  3<span style="text-decoration: underline;">423</span>522 3<span style="text-decoration: underline;">423</span>834 CpG: 28     312     28   215   17.9  68.9   0.87</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span>   612 chr7  3<span style="text-decoration: underline;">617</span>298 3<span style="text-decoration: underline;">617</span>534 CpG: 17     236     17   146   14.4  61.9   0.76</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span>   612 chr7  3<span style="text-decoration: underline;">629</span>691 3<span style="text-decoration: underline;">630</span>056 CpG: 32     365     32   195   17.5  53.4   1.24</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span>   612 chr7  3<span style="text-decoration: underline;">645</span>002 3<span style="text-decoration: underline;">645</span>963 CpG: 106    961    106   652   22.1  67.8   0.96</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span>   612 chr7  3<span style="text-decoration: underline;">665</span>238 3<span style="text-decoration: underline;">665</span>517 CpG: 27     279     27   184   19.4  65.9   0.93</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1,183 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3 more variables: transcript_id &lt;chr&gt;, strand &lt;chr&gt;, symbol &lt;chr&gt;</span></span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_agg_regions.html" class="external-link">plot_agg_regions</a></span><span class="op">(</span><span class="va">nmr</span>, <span class="va">cgi_anno</span>, group_col <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="workflow_files/figure-html/fig4-cgi-agg-1.png" alt="**Aggregated CpG island-level plot**. The mean methylation proportion is shown across CpG islands on chromosome 7, aggregated by experimental group. Methylation proportions are individual CpG sites are calculated along relative genomic positions, with fixed flanking regions, then averaged across all CpG islands in each experimental group to produce a smoothed trendline." width="600"><p class="caption">
<strong>Aggregated CpG island-level plot</strong>. The mean methylation
proportion is shown across CpG islands on chromosome 7, aggregated by
experimental group. Methylation proportions are individual CpG sites are
calculated along relative genomic positions, with fixed flanking
regions, then averaged across all CpG islands in each experimental group
to produce a smoothed trendline.
</p>
</div>
<p>As expected, we see a clear pattern of demethylation over CpG islands
on chromosome 7, relative to neighbouring genomic regions.</p>
</div>
<div class="section level2">
<h2 id="differential-methylation">Differential methylation<a class="anchor" aria-label="anchor" href="#differential-methylation"></a>
</h2>
<p><code>NanoMethViz</code> is primarily designed for visualising
methylation data. In order to perform differential methylation analysis,
we can use other Bioconductor packages that implement sophisticated
statistical procedures, such as <code>dmrseq</code>. The
<code><a href="https://rdrr.io/pkg/dmrseq/man/dmrseq.html" class="external-link">dmrseq()</a></code> function requires a <code>BSseq</code> object as
input, so we convert the <code>NanoMethResult</code> object to a
<code>BSseq</code> object using the <code><a href="https://rdrr.io/pkg/NanoMethViz/man/methy_to_bsseq.html" class="external-link">methy_to_bsseq()</a></code>
function.</p>
<p>We remove any sites that are zero in all of the samples of either
experimental group; these low coverage regions are generally not
informative for differential methylation analysis. The
<code>BSseq</code> object stores the methylation data as two matrices,
the methylation matrix and the coverage matrix (Figure @ref(fig:fig1)C),
with the option to store additional information such as sample
annotation and genomic regions.</p>
<p>This conversion process usually takes some time, so we will load a
pre-generated <code>BSseq</code> object. The code used to generate the
object is provided below.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"data/bss.rds"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">bss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/methy_to_bsseq.html" class="external-link">methy_to_bsseq</a></span><span class="op">(</span><span class="va">nmr</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">bss</span>, <span class="st">"data/bss.rds"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">bss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"data/bss.rds"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">pat_cov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bsseq/man/getCoverage.html" class="external-link">getCoverage</a></span><span class="op">(</span><span class="va">bss</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">bss</span><span class="op">$</span><span class="va">group</span> <span class="op">==</span> <span class="st">"nsc_pat"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">mat_cov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bsseq/man/getCoverage.html" class="external-link">getCoverage</a></span><span class="op">(</span><span class="va">bss</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">bss</span><span class="op">$</span><span class="va">group</span> <span class="op">==</span> <span class="st">"nsc_mat"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">low_cov</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">pat_cov</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">3</span><span class="op">)</span> <span class="op">|</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">mat_cov</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/proportions.html" class="external-link">proportions</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">low_cov</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## low_cov</span></span>
<span><span class="co">##     FALSE      TRUE </span></span>
<span><span class="co">## 0.2853227 0.7146773</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bss</span> <span class="op">&lt;-</span> <span class="va">bss</span><span class="op">[</span><span class="op">!</span><span class="va">low_cov</span>, <span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="mds-plot">MDS Plot<a class="anchor" aria-label="anchor" href="#mds-plot"></a>
</h2>
<p>Using the <code>BSseq</code> object, we are able to create a
multi-dimensional scaling (MDS) plot to visualise the relative
differences in methylation between the samples, a method commonly used
in differential gene expression analysis. The MDS function within
<code>NanoMethViz</code> uses the <code>plotMDS()</code> function from
<code>limma</code> <span class="citation">(16)</span>. Since
<code>BSseq</code> objects store the methylation data as two matrices,
the methylation matrix and the coverage matrix, we need to convert the
<code>BSseq</code> object such that each CpG site from each sample is
represented by a single value in a matrix. This can be done using the
<code><a href="https://rdrr.io/pkg/NanoMethViz/man/bsseq_to_log_methy_ratio.html" class="external-link">bsseq_to_log_methy_ratio()</a></code> function, which converts the
<code>BSseq</code> object to a matrix of the log of the methylation
ratio with a small count added to prevent division by zero.</p>
<p>The <code><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_mds.html" class="external-link">plot_mds()</a></code> function can then be used to create an
MDS plot of the samples using the log methylation ratio matrix as input.
The samples are coloured by group – in this case, maternal and
paternal.</p>
<p>Because the number of CpG sites is large, here we aggregate the
methylation data over CpG islands to both reduce the size of the data
and produce a more stable measure of methylation. This is done by
providing the <code>regions = cgi_anno</code> argument to
<code><a href="https://rdrr.io/pkg/NanoMethViz/man/bsseq_to_log_methy_ratio.html" class="external-link">bsseq_to_log_methy_ratio()</a></code>. Without the argument, the
function will use individual CpG sites. After the data is aggregated,
<code><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_mds.html" class="external-link">plot_mds()</a></code> uses only the top 500 most variable features to
create the MDS plot (Figure @ref(fig:fig5-mds)).</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lmr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/bsseq_to_log_methy_ratio.html" class="external-link">bsseq_to_log_methy_ratio</a></span><span class="op">(</span></span>
<span>    <span class="va">bss</span>,</span>
<span>    regions <span class="op">=</span> <span class="va">cgi_anno</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_mds.html" class="external-link">plot_mds</a></span><span class="op">(</span><span class="va">lmr</span>, groups <span class="op">=</span> <span class="va">bss</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="workflow_files/figure-html/fig5-mds-1.png" alt="**MDS plot**, The multi-dimensional plot is created based on the top 500 most variable genes in on chromosome 7 in terms of log-methylation-ratio over the gene bodies. The samples are coloured by group, and distances between samples is reflective of the difference in the methylation profile across genes in each sample." width="600"><p class="caption">
<strong>MDS plot</strong>, The multi-dimensional plot is created based
on the top 500 most variable genes in on chromosome 7 in terms of
log-methylation-ratio over the gene bodies. The samples are coloured by
group, and distances between samples is reflective of the difference in
the methylation profile across genes in each sample.
</p>
</div>
<p>From this MDS, we see that there is clear separation between the
maternal and paternal groups based on CpG islands on chromosome 7,
suggesting that the most variable CpG islands are differentially
methylated between the two groups. Given that this is autosomal data,
the separation is likely due to imprinting or strain-specific effects.
We therefore expect to find some differentially methylated regions
between the maternal and paternal groups.</p>
</div>
<div class="section level2">
<h2 id="differentially-methylated-regions">Differentially methylated regions<a class="anchor" aria-label="anchor" href="#differentially-methylated-regions"></a>
</h2>
<p>Another common analysis is to identify differentially methylated
regions (DMRs) between experimental groups. It is often informative to
identify DMRs rather than just single differentially-methylated CpG
sites, as these may correspond to gene regulatory regions such as
promoters, silencers and enhancers, and have biological relevance for
differential epigenetic regulation. However, DMRs are not easy to
identify, as DMR calling requires the identification of individual
significant sites, some determination of whether nearby sites should be
aggregated into a region, and a statistical test to determine if the
region as a whole is differentially methylated.</p>
<p>The <code>dmrseq</code> package can be used to identify DMRs between
the two groups in our data. The <code><a href="https://rdrr.io/pkg/dmrseq/man/dmrseq.html" class="external-link">dmrseq()</a></code> function from the
package requires a <code>BSseq</code> object as input, along with the
name of the covariate to test for differential methylation. We will test
for differential methylation between the maternal and paternal groups.
The <code><a href="https://rdrr.io/pkg/dmrseq/man/dmrseq.html" class="external-link">dmrseq()</a></code> function returns a <code>GRanges</code> object
containing the DMRs identified by the function. The <code>GRanges</code>
object contains the genomic coordinates of the DMRs, along with
information about the statistical significance of the DMRs.</p>
<p>There are alternatives to <code>dmrseq</code> for identifying DMRs,
such as <code>bsseq</code>, <code>DSS</code> and <code>edgeR</code>. For
<code>bsseq</code> and <code>DSS</code>, the same <code>BSseq</code>
object can be used as input, while <code>edgeR</code> requires a
<code>DGEList</code> object as input (Figure @ref(fig:fig1)C). The
<code>DGEList</code> object can be created from the <code>BSseq</code>
object using the <code><a href="https://rdrr.io/pkg/NanoMethViz/man/bsseq_to_edger.html" class="external-link">bsseq_to_edger()</a></code> function, or from the
<code>NanoMethResult</code> object using the
<code><a href="https://rdrr.io/pkg/NanoMethViz/man/methy_to_edger.html" class="external-link">methy_to_edger()</a></code> function (Figure @ref(fig:fig1)B).</p>
<p><code>bsseq</code>, <code>DSS</code> and <code>dmrseq</code> all use
a similar approach to identify DMRs: first, they use a statistical test
to compare the methylation levels between the two groups at each CpG
site, followed by a clustering step to identify regions of the genome
with consistent differences in methylation levels for de novo DMR
discovery. On the other hand, <code>edgeR</code> requires users to
choose the regions to test for differential methylation, which is
reflected by the regions over which they choose to summarise counts, and
therefore cannot be used for de novo DMR discovery. <code>dmrseq</code>
and <code>edgeR</code> produce
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values
for each region, while <code>bsseq</code> and <code>DSS</code> only
provide
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values
for each CpG site along with an aggregated area statistic.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"data/regions.rds"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu">dmrseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dmrseq/man/dmrseq.html" class="external-link">dmrseq</a></span><span class="op">(</span><span class="va">bss</span>, testCovariate <span class="op">=</span> <span class="st">"group"</span>, minNumRegion <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">regions</span>, <span class="st">"data/regions.rds"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"data/regions.rds"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>A number of regions will be produced by <code>dmrseq</code>,
including regions that were aggregated because they contained
significant CpG sites but subsequently failed to reach statistical
significance at a region level. It is difficult to directly interpret
the regions returned by differential methylation analysis, and we
commonly wish to identify these regions are possible control regions for
genes. We do this by associating DMRs with genes using proximity to the
gene transcription start site (TSS), first by constructing an annotation
of a promoter region, then overlapping it with the DMRs. We use a
liberal window of 10kb around the TSS to capture potential promoter
regions, this is done by transforming the gene annotation to a
<code>GRanges</code> object, then using the <code><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">promoters()</a></code>
function from <code>IRanges</code> <span class="citation">(17)</span> to
define the promoter region as 5kb upstream and 5kb downstream of the
TSS. We then use the <code><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_intersect()</a></code> function from
<code>plyranges</code> to find the DMRs that overlap with the gene TSS
regions. This will return a <code>GRanges</code> object containing the
regions where DMRs which overlap with the promoter region, and the
associated gene in that region.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_anno_gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">gene_anno</span>, <span class="st">"GRanges"</span><span class="op">)</span></span>
<span><span class="va">gene_anno_gr_tss</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">promoters</a></span><span class="op">(</span><span class="va">gene_anno_gr</span>, upstream <span class="op">=</span> <span class="fl">5000</span>, downstream <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gene_dmr_overlaps</span> <span class="op">&lt;-</span> <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_intersect</a></span><span class="op">(</span><span class="va">regions</span>, <span class="va">gene_anno_gr_tss</span><span class="op">)</span></span>
<span><span class="va">gene_dmr_overlaps</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 868 ranges and 9 metadata columns:</span></span>
<span><span class="co">##         seqnames              ranges strand |         L      area       beta</span></span>
<span><span class="co">##            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">##     [1]     chr7     6727308-6731838      * |       157  128.8218   -1.84082</span></span>
<span><span class="co">##     [2]     chr7     6727308-6731838      * |       157  128.8218   -1.84082</span></span>
<span><span class="co">##     [3]     chr7 143294268-143296757      * |       112   92.9051   -1.87215</span></span>
<span><span class="co">##     [4]     chr7   60003279-60005228      * |        60   50.9278   -1.82583</span></span>
<span><span class="co">##     [5]     chr7   60003279-60005228      * |        60   50.9278   -1.82583</span></span>
<span><span class="co">##     ...      ...                 ...    ... .       ...       ...        ...</span></span>
<span><span class="co">##   [864]     chr7 126579062-126580889      * |        27   5.00269  0.0232921</span></span>
<span><span class="co">##   [865]     chr7 126580818-126580889      * |        27   5.00269  0.0232921</span></span>
<span><span class="co">##   [866]     chr7 126579942-126580889      * |        27   5.00269  0.0232921</span></span>
<span><span class="co">##   [867]     chr7   19696925-19697636      * |        23   2.65049 -0.0118438</span></span>
<span><span class="co">##   [868]     chr7   19696925-19697636      * |        23   2.65049 -0.0118438</span></span>
<span><span class="co">##               stat        pval      qval         index     gene_id      symbol</span></span>
<span><span class="co">##          &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;     &lt;IRanges&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##     [1]   -40.6010 5.71657e-05 0.0401684   27111-27267       18616        Peg3</span></span>
<span><span class="co">##     [2]   -40.6010 5.71657e-05 0.0401684   27111-27267       57775       Usp29</span></span>
<span><span class="co">##     [3]   -32.5253 1.14331e-04 0.0401684 910828-910939       63830    Kcnq1ot1</span></span>
<span><span class="co">##     [4]   -20.1018 1.14331e-04 0.0401684 335287-335346       52480      Snhg14</span></span>
<span><span class="co">##     [5]   -20.1018 1.14331e-04 0.0401684 335287-335346       84704       Snurf</span></span>
<span><span class="co">##     ...        ...         ...       ...           ...         ...         ...</span></span>
<span><span class="co">##   [864]  0.0925951    0.996341  0.997876 764165-764191   102465641     Mir7059</span></span>
<span><span class="co">##   [865]  0.0925951    0.996341  0.997876 764165-764191       12752        Cln3</span></span>
<span><span class="co">##   [866]  0.0925951    0.996341  0.997876 764165-764191      171504       Apobr</span></span>
<span><span class="co">##   [867] -0.0795915    0.996456  0.997876   81437-81459       11812       Apoc1</span></span>
<span><span class="co">##   [868] -0.0795915    0.996456  0.997876   81437-81459       11816        Apoe</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>After we identify the DMRs that overlap with the gene TSS regions, we
can filter for the significant DMRs using the qval column. This is the
adjusted
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-value
for the DMRs, calculated using the Benjamini-Hochberg procedure to
control false-discovery rates for multiple-testing. We can then select
the columns we are interested in, such as the gene symbol, chromosome,
start and end positions of the DMRs, the strand of the gene, and the
q-value of the DMRs.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dmr_regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">gene_dmr_overlaps</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="st">"seqnames"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">signif_regions</span> <span class="op">&lt;-</span> <span class="va">dmr_regions</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">qval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="va">signif_regions</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">symbol</span>, <span class="va">chr</span>, <span class="va">start</span>, <span class="va">end</span>, <span class="va">strand</span>, <span class="va">qval</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 8 × 6</span></span></span>
<span><span class="co">##   symbol   chr       start       end strand   qval</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> Peg3     chr7    6<span style="text-decoration: underline;">727</span>308   6<span style="text-decoration: underline;">731</span>838 *      0.040<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> Usp29    chr7    6<span style="text-decoration: underline;">727</span>308   6<span style="text-decoration: underline;">731</span>838 *      0.040<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> Kcnq1ot1 chr7  143<span style="text-decoration: underline;">294</span>268 143<span style="text-decoration: underline;">296</span>757 *      0.040<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> Snhg14   chr7   60<span style="text-decoration: underline;">003</span>279  60<span style="text-decoration: underline;">005</span>228 *      0.040<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> Snurf    chr7   60<span style="text-decoration: underline;">003</span>279  60<span style="text-decoration: underline;">005</span>228 *      0.040<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> Mkrn3    chr7   62<span style="text-decoration: underline;">416</span>953  62<span style="text-decoration: underline;">423</span>032 *      0.040<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">7</span> Cdkn1c   chr7  143<span style="text-decoration: underline;">459</span>739 143<span style="text-decoration: underline;">462</span>038 *      0.040<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">8</span> Peg12    chr7   62<span style="text-decoration: underline;">461</span>469  62<span style="text-decoration: underline;">464</span>053 *      0.040<span style="text-decoration: underline;">2</span></span></span></code></pre>
<p>We now have a list of DMRs which overlap the TSS of genes on
chromosome 7 and have a q-value &lt; 0.05. The genes on this list are
all known imprinted genes in the mouse, including <em>Peg3</em>, which
we plotted in the first section of this workflow. We can plot the
methylations of these genes and highlight the DMRs by setting the
<code><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene()</a></code> function’s <code>anno_regions</code> argument
to the regions identified. This will highlight the significant regions
in the gene plot in a shaded band. Here we will plot <em>Peg3</em>,
<em>Kcnq1ot1</em>, and <em>Cdkn1c</em> (Figure
@ref(fig:fig6-dmr-genes)). We can see that for all three of these genes,
as required by our filtering criteria, the differentially methylated
regions fall near the TSS.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"NanoMethViz.highlight_col"</span> <span class="op">=</span> <span class="st">"green"</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_elements.html" class="external-link">wrap_elements</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene</a></span><span class="op">(</span><span class="va">nmr</span>, <span class="st">"Peg3"</span>, anno_regions <span class="op">=</span> <span class="va">signif_regions</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span></span>
<span> <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_elements.html" class="external-link">wrap_elements</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene</a></span><span class="op">(</span><span class="va">nmr</span>, <span class="st">"Kcnq1ot1"</span>, anno_regions <span class="op">=</span> <span class="va">signif_regions</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span></span>
<span> <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_elements.html" class="external-link">wrap_elements</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene</a></span><span class="op">(</span><span class="va">nmr</span>, <span class="st">"Cdkn1c"</span>, anno_regions <span class="op">=</span> <span class="va">signif_regions</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="workflow_files/figure-html/fig6-dmr-genes-1.png" alt="**Gene plots with DMR**. The plots show the genes *Peg3*, *Kcnq1ot1*, and *Cdkn1c* along with green highlighted region denoting the DMR discovered by *dmrseq*." width="600"><p class="caption">
<strong>Gene plots with DMR</strong>. The plots show the genes
<em>Peg3</em>, <em>Kcnq1ot1</em>, and <em>Cdkn1c</em> along with green
highlighted region denoting the DMR discovered by <em>dmrseq</em>.
</p>
</div>
<p>We can see that the DMR for <em>Cdkn1c</em> extends beyond the region
that is plotted when we use <code><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene()</a></code>. We can plot these
DMRs using <code><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_region.html" class="external-link">plot_region()</a></code>, with 2000 bases of flanking
sequence added to both sides of the DMR to better visualise the full
scope of the differentially methylated region (Figure
@ref(fig:fig7-cdkn1-expand)).</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_region.html" class="external-link">plot_region</a></span><span class="op">(</span><span class="va">nmr</span>, <span class="st">"chr7"</span>, <span class="fl">143459739</span><span class="op">-</span><span class="fl">2000</span>, <span class="fl">143462038</span><span class="op">+</span><span class="fl">2000</span>, anno_regions <span class="op">=</span> <span class="va">signif_regions</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="workflow_files/figure-html/fig7-cdkn1-expand-1.png" alt="**Expanded region around *Cdkn1c***. This plot shows a more expanded region around the gene *Cdkn1c* to show the full DMR." width="600"><p class="caption">
<strong>Expanded region around <em>Cdkn1c</em></strong>. This plot shows
a more expanded region around the gene <em>Cdkn1c</em> to show the full
DMR.
</p>
</div>
<p>We have now successfully used <code>dmrseq</code> to statistically
identify differentially methylated regions around known imprinted genes.
Additionally, we have used that information to visualise the methylation
data of the genes associated with the DMRs, and identify maternal- and
paternal-specific methylation of the TSS.</p>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>In this workflow, we have demonstrated a complete end-to-end pipeline
for analysing and visualising methylation data generated by ONT
long-read sequencing using <code>NanoMethViz</code>, <code>dmrseq</code>
and other Bioconductor packages. We have shown how to preprocess the
data, load it into <code>NanoMethViz</code>, visualise the methylation
patterns of genes and aggregated features, cluster samples using MDS
plots, and identify DMRs between experimental groups. We have also shown
how to associate DMRs with genes and visualise the methylation patterns
of genes associated with DMRs. While the DMRs identified in this
tutorial were all associated with known imprinted genes, this workflow
could be used in other experimental contexts to identify candidate genes
involved in disease pathogenesis, or discover new gene regulatory
elements. Further analysis could include enrichment analysis of the
differentially-methylated genes, to bring out pathways that may be up-
or downregulated in different processes. Ultimately, this workflow
demonstrates a toolkit for the exploration and analysis of methylation
data from long-read sequencing, which can be applied to a range of
resolution levels and biological contexts. While the workflow was
demonstrated on 5mC ONT nanopore data, it can also be applied to other
long-read DNA methylation data such as that produced by PacBio SMRT
platforms as long as the data is provided in a modBAM format. In
principle this can be applied to other modification types such as 5hmC
or 6mA, as well as potentially RNA modifications, but this is not be
thoroughly tested.</p>
</div>
<div class="section level2">
<h2 id="software-avaliability">Software Avaliability<a class="anchor" aria-label="anchor" href="#software-avaliability"></a>
</h2>
<p>This workflow makes use of various R packages available from the
Bioconductor project, with version numbers shown below. The
<code>LongMod123</code> workflow package itself is available at and from
Bioconductor.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Etc/UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] NanoMethViz_3.1.5           ggplot2_3.5.1              </span></span>
<span><span class="co">##  [3] plyranges_1.25.0            readr_2.1.5                </span></span>
<span><span class="co">##  [5] dplyr_1.1.4                 dmrseq_1.25.1              </span></span>
<span><span class="co">##  [7] bsseq_1.41.0                SummarizedExperiment_1.35.1</span></span>
<span><span class="co">##  [9] Biobase_2.65.1              MatrixGenerics_1.17.0      </span></span>
<span><span class="co">## [11] matrixStats_1.3.0           GenomicRanges_1.57.1       </span></span>
<span><span class="co">## [13] GenomeInfoDb_1.41.1         IRanges_2.39.2             </span></span>
<span><span class="co">## [15] S4Vectors_0.43.2            BiocGenerics_0.51.0        </span></span>
<span><span class="co">## [17] BiocFileCache_2.13.0        dbplyr_2.5.0               </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] splines_4.4.1                            </span></span>
<span><span class="co">##   [2] BiocIO_1.15.2                            </span></span>
<span><span class="co">##   [3] bitops_1.0-8                             </span></span>
<span><span class="co">##   [4] filelock_1.0.3                           </span></span>
<span><span class="co">##   [5] tibble_3.2.1                             </span></span>
<span><span class="co">##   [6] R.oo_1.26.0                              </span></span>
<span><span class="co">##   [7] XML_3.99-0.17                            </span></span>
<span><span class="co">##   [8] lifecycle_1.0.4                          </span></span>
<span><span class="co">##   [9] vroom_1.6.5                              </span></span>
<span><span class="co">##  [10] lattice_0.22-6                           </span></span>
<span><span class="co">##  [11] magrittr_2.0.3                           </span></span>
<span><span class="co">##  [12] limma_3.61.9                             </span></span>
<span><span class="co">##  [13] sass_0.4.9                               </span></span>
<span><span class="co">##  [14] rmarkdown_2.28                           </span></span>
<span><span class="co">##  [15] jquerylib_0.1.4                          </span></span>
<span><span class="co">##  [16] yaml_2.3.10                              </span></span>
<span><span class="co">##  [17] doRNG_1.8.6                              </span></span>
<span><span class="co">##  [18] DBI_1.2.3                                </span></span>
<span><span class="co">##  [19] RColorBrewer_1.1-3                       </span></span>
<span><span class="co">##  [20] abind_1.4-5                              </span></span>
<span><span class="co">##  [21] zlibbioc_1.51.1                          </span></span>
<span><span class="co">##  [22] purrr_1.0.2                              </span></span>
<span><span class="co">##  [23] R.utils_2.12.3                           </span></span>
<span><span class="co">##  [24] RCurl_1.98-1.16                          </span></span>
<span><span class="co">##  [25] rappdirs_0.3.3                           </span></span>
<span><span class="co">##  [26] cpp11_0.5.0                              </span></span>
<span><span class="co">##  [27] GenomeInfoDbData_1.2.12                  </span></span>
<span><span class="co">##  [28] irlba_2.3.5.1                            </span></span>
<span><span class="co">##  [29] pkgdown_2.1.0                            </span></span>
<span><span class="co">##  [30] permute_0.9-7                            </span></span>
<span><span class="co">##  [31] DelayedMatrixStats_1.27.3                </span></span>
<span><span class="co">##  [32] codetools_0.2-20                         </span></span>
<span><span class="co">##  [33] DelayedArray_0.31.11                     </span></span>
<span><span class="co">##  [34] tidyselect_1.2.1                         </span></span>
<span><span class="co">##  [35] outliers_0.15                            </span></span>
<span><span class="co">##  [36] farver_2.1.2                             </span></span>
<span><span class="co">##  [37] UCSC.utils_1.1.0                         </span></span>
<span><span class="co">##  [38] ScaledMatrix_1.13.0                      </span></span>
<span><span class="co">##  [39] GenomicAlignments_1.41.0                 </span></span>
<span><span class="co">##  [40] jsonlite_1.8.8                           </span></span>
<span><span class="co">##  [41] annotatr_1.31.0                          </span></span>
<span><span class="co">##  [42] e1071_1.7-14                             </span></span>
<span><span class="co">##  [43] iterators_1.0.14                         </span></span>
<span><span class="co">##  [44] systemfonts_1.1.0                        </span></span>
<span><span class="co">##  [45] foreach_1.5.2                            </span></span>
<span><span class="co">##  [46] dbscan_1.2-0                             </span></span>
<span><span class="co">##  [47] tools_4.4.1                              </span></span>
<span><span class="co">##  [48] ragg_1.3.2                               </span></span>
<span><span class="co">##  [49] Rcpp_1.0.13                              </span></span>
<span><span class="co">##  [50] glue_1.7.0                               </span></span>
<span><span class="co">##  [51] SparseArray_1.5.31                       </span></span>
<span><span class="co">##  [52] mgcv_1.9-1                               </span></span>
<span><span class="co">##  [53] xfun_0.47                                </span></span>
<span><span class="co">##  [54] HDF5Array_1.33.6                         </span></span>
<span><span class="co">##  [55] withr_3.0.1                              </span></span>
<span><span class="co">##  [56] BiocManager_1.30.25                      </span></span>
<span><span class="co">##  [57] fastmap_1.2.0                            </span></span>
<span><span class="co">##  [58] rhdf5filters_1.17.0                      </span></span>
<span><span class="co">##  [59] fansi_1.0.6                              </span></span>
<span><span class="co">##  [60] digest_0.6.37                            </span></span>
<span><span class="co">##  [61] rsvd_1.0.5                               </span></span>
<span><span class="co">##  [62] R6_2.5.1                                 </span></span>
<span><span class="co">##  [63] textshaping_0.4.0                        </span></span>
<span><span class="co">##  [64] colorspace_2.1-1                         </span></span>
<span><span class="co">##  [65] Cairo_1.6-2                              </span></span>
<span><span class="co">##  [66] gtools_3.9.5                             </span></span>
<span><span class="co">##  [67] RSQLite_2.3.7                            </span></span>
<span><span class="co">##  [68] R.methodsS3_1.8.2                        </span></span>
<span><span class="co">##  [69] utf8_1.2.4                               </span></span>
<span><span class="co">##  [70] tidyr_1.3.1                              </span></span>
<span><span class="co">##  [71] generics_0.1.3                           </span></span>
<span><span class="co">##  [72] data.table_1.16.0                        </span></span>
<span><span class="co">##  [73] rtracklayer_1.65.0                       </span></span>
<span><span class="co">##  [74] class_7.3-22                             </span></span>
<span><span class="co">##  [75] httr_1.4.7                               </span></span>
<span><span class="co">##  [76] htmlwidgets_1.6.4                        </span></span>
<span><span class="co">##  [77] S4Arrays_1.5.7                           </span></span>
<span><span class="co">##  [78] org.Mm.eg.db_3.19.1                      </span></span>
<span><span class="co">##  [79] regioneR_1.37.0                          </span></span>
<span><span class="co">##  [80] pkgconfig_2.0.3                          </span></span>
<span><span class="co">##  [81] gtable_0.3.5                             </span></span>
<span><span class="co">##  [82] blob_1.2.4                               </span></span>
<span><span class="co">##  [83] XVector_0.45.0                           </span></span>
<span><span class="co">##  [84] htmltools_0.5.8.1                        </span></span>
<span><span class="co">##  [85] scales_1.3.0                             </span></span>
<span><span class="co">##  [86] png_0.1-8                                </span></span>
<span><span class="co">##  [87] knitr_1.48                               </span></span>
<span><span class="co">##  [88] tzdb_0.4.0                               </span></span>
<span><span class="co">##  [89] reshape2_1.4.4                           </span></span>
<span><span class="co">##  [90] rjson_0.2.22                             </span></span>
<span><span class="co">##  [91] nlme_3.1-166                             </span></span>
<span><span class="co">##  [92] curl_5.2.2                               </span></span>
<span><span class="co">##  [93] bumphunter_1.47.0                        </span></span>
<span><span class="co">##  [94] proxy_0.4-27                             </span></span>
<span><span class="co">##  [95] cachem_1.1.0                             </span></span>
<span><span class="co">##  [96] rhdf5_2.49.0                             </span></span>
<span><span class="co">##  [97] stringr_1.5.1                            </span></span>
<span><span class="co">##  [98] BiocVersion_3.20.0                       </span></span>
<span><span class="co">##  [99] vipor_0.4.7                              </span></span>
<span><span class="co">## [100] parallel_4.4.1                           </span></span>
<span><span class="co">## [101] AnnotationDbi_1.67.0                     </span></span>
<span><span class="co">## [102] ggrastr_1.0.2                            </span></span>
<span><span class="co">## [103] restfulr_0.0.15                          </span></span>
<span><span class="co">## [104] desc_1.4.3                               </span></span>
<span><span class="co">## [105] pillar_1.9.0                             </span></span>
<span><span class="co">## [106] grid_4.4.1                               </span></span>
<span><span class="co">## [107] vctrs_0.6.5                              </span></span>
<span><span class="co">## [108] BiocSingular_1.21.2                      </span></span>
<span><span class="co">## [109] beachmat_2.21.5                          </span></span>
<span><span class="co">## [110] beeswarm_0.4.0                           </span></span>
<span><span class="co">## [111] evaluate_0.24.0                          </span></span>
<span><span class="co">## [112] GenomicFeatures_1.57.0                   </span></span>
<span><span class="co">## [113] cli_3.6.3                                </span></span>
<span><span class="co">## [114] locfit_1.5-9.10                          </span></span>
<span><span class="co">## [115] compiler_4.4.1                           </span></span>
<span><span class="co">## [116] Rsamtools_2.21.1                         </span></span>
<span><span class="co">## [117] rlang_1.1.4                              </span></span>
<span><span class="co">## [118] crayon_1.5.3                             </span></span>
<span><span class="co">## [119] rngtools_1.5.2                           </span></span>
<span><span class="co">## [120] labeling_0.4.3                           </span></span>
<span><span class="co">## [121] ggbeeswarm_0.7.2                         </span></span>
<span><span class="co">## [122] plyr_1.8.9                               </span></span>
<span><span class="co">## [123] forcats_1.0.0                            </span></span>
<span><span class="co">## [124] fs_1.6.4                                 </span></span>
<span><span class="co">## [125] TxDb.Mmusculus.UCSC.mm10.knownGene_3.10.0</span></span>
<span><span class="co">## [126] stringi_1.8.4                            </span></span>
<span><span class="co">## [127] BiocParallel_1.39.0                      </span></span>
<span><span class="co">## [128] assertthat_0.2.1                         </span></span>
<span><span class="co">## [129] munsell_0.5.1                            </span></span>
<span><span class="co">## [130] Biostrings_2.73.1                        </span></span>
<span><span class="co">## [131] Matrix_1.7-0                             </span></span>
<span><span class="co">## [132] BSgenome_1.73.0                          </span></span>
<span><span class="co">## [133] hms_1.1.3                                </span></span>
<span><span class="co">## [134] patchwork_1.2.0                          </span></span>
<span><span class="co">## [135] sparseMatrixStats_1.17.2                 </span></span>
<span><span class="co">## [136] bit64_4.0.5                              </span></span>
<span><span class="co">## [137] Rhdf5lib_1.27.0                          </span></span>
<span><span class="co">## [138] KEGGREST_1.45.1                          </span></span>
<span><span class="co">## [139] statmod_1.5.0                            </span></span>
<span><span class="co">## [140] highr_0.11                               </span></span>
<span><span class="co">## [141] AnnotationHub_3.13.3                     </span></span>
<span><span class="co">## [142] memoise_2.0.1                            </span></span>
<span><span class="co">## [143] bslib_0.8.0                              </span></span>
<span><span class="co">## [144] bit_4.0.5</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="acknowledgements">Acknowledgements<a class="anchor" aria-label="anchor" href="#acknowledgements"></a>
</h2>
<p>We thank Kathleen Zeglinski for creating the <code>NanoMethViz</code>
logo. This work was supported by Australian National Health and Medical
Research Council (NHMRC) Investigator Grants (GNT2007996 to Q.G.,
GNT1194345 to M.E.B. and GNT2017257 to M.E.R), Medical Research Future
Fund Researcher Exchange and Development in Industry Fellowship
(REDIF249 to M.E.R), Victorian State Government Operational
Infrastructure Support and Australian Government NHMRC IRIISS.</p>
</div>
<div class="section level2">
<h2 id="author-contribution">Author Contribution<a class="anchor" aria-label="anchor" href="#author-contribution"></a>
</h2>
<p>S.S, L.X. and J.L. developed the workflow, performed data analysis,
generated figures and wrote the manuscript. T.C., K.B. and M.E.B.
generated data analysed in the workflow. P.F.H., M.E.B, Q.G. and M.E.R.
supervised the project. All authors read and approved the final
manuscript.</p>
</div>
<div class="section level2">
<h2 id="competing-interests">Competing Interests<a class="anchor" aria-label="anchor" href="#competing-interests"></a>
</h2>
<p>The authors declare that they have no competing interests.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body">
<div id="ref-jinDNAMethylationSuperior2011" class="csl-entry">
<div class="csl-left-margin">1. </div>
<div class="csl-right-inline">Jin
B, Li Y, Robertson KD. <a href="https://doi.org/10.1177/1947601910393957" class="external-link"><span>DNA
Methylation</span>: <span>Superior</span> or <span>Subordinate</span> in
the <span>Epigenetic Hierarchy</span>?</a> Genes &amp; Cancer. 2011
Jun;2(6):607–17. </div>
</div>
<div id="ref-robertsonDNAMethylationHuman2005" class="csl-entry">
<div class="csl-left-margin">2. </div>
<div class="csl-right-inline">Robertson KD. <a href="https://doi.org/10.1038/nrg1655" class="external-link"><span>DNA</span> methylation and
human disease</a>. Nature Reviews Genetics. 2005 Aug;6(8):597–610.
</div>
</div>
<div id="ref-frommerGenomicSequencingProtocol1992" class="csl-entry">
<div class="csl-left-margin">3. </div>
<div class="csl-right-inline">Frommer M, McDonald LE, Millar DS, Collis CM,
Watt F, Grigg GW, et al. <a href="https://doi.org/10.1073/pnas.89.5.1827" class="external-link">A genomic sequencing
protocol that yields a positive display of 5-methylcytosine residues in
individual <span>DNA</span> strands</a>. Proceedings of the National
Academy of Sciences of the United States of America. 1992
Mar;89(5):1827–31. </div>
</div>
<div id="ref-pidsleyCriticalEvaluationIllumina2016" class="csl-entry">
<div class="csl-left-margin">4. </div>
<div class="csl-right-inline">Pidsley R, Zotenko E, Peters TJ, Lawrence MG,
Risbridger GP, Molloy P, et al. <a href="https://doi.org/10.1186/s13059-016-1066-1" class="external-link">Critical evaluation of
the <span>Illumina MethylationEPIC BeadChip</span> microarray for
whole-genome <span>DNA</span> methylation profiling</a>. Genome Biology.
2016 Oct;17(1):208. </div>
</div>
<div id="ref-gouilLatestTechniquesStudy2019" class="csl-entry">
<div class="csl-left-margin">5. </div>
<div class="csl-right-inline">Gouil Q, Keniry A. <a href="https://doi.org/10.1042/EBC20190027" class="external-link">Latest techniques to study
<span>DNA</span> methylation</a>. Essays in Biochemistry. 2019
Dec;63(6):639–48. </div>
</div>
<div id="ref-liuDNAMethylationAtlas2021a" class="csl-entry">
<div class="csl-left-margin">6. </div>
<div class="csl-right-inline">Liu
H, Zhou J, Tian W, Luo C, Bartlett A, Aldridge A, et al. <a href="https://doi.org/10.1038/s41586-020-03182-8" class="external-link"><span>DNA</span>
methylation atlas of the mouse brain at single-cell resolution</a>.
Nature. 2021 Oct;598(7879):120–8. </div>
</div>
<div id="ref-suNanoMethVizBioconductorPackage2021a" class="csl-entry">
<div class="csl-left-margin">7. </div>
<div class="csl-right-inline">Su
S, Gouil Q, Blewitt ME, Cook D, Hickey PF, Ritchie ME. <a href="https://doi.org/10.1371/journal.pcbi.1009524" class="external-link"><span>NanoMethViz</span>:
<span>An R</span>/<span>Bioconductor</span> package for visualizing
long-read methylation data</a>. PLOS Computational Biology. 2021
Oct;17(10):e1009524. </div>
</div>
<div id="ref-hansenBSmoothWholeGenome2012" class="csl-entry">
<div class="csl-left-margin">8. </div>
<div class="csl-right-inline">Hansen KD, Langmead B, Irizarry RA. <a href="https://doi.org/10.1186/gb-2012-13-10-r83" class="external-link"><span>BSmooth</span>:
From whole genome bisulfite sequencing reads to differentially
methylated regions</a>. Genome Biology. 2012 Oct;13(10):R83. </div>
</div>
<div id="ref-parkDifferentialMethylationAnalysis2016" class="csl-entry">
<div class="csl-left-margin">9. </div>
<div class="csl-right-inline">Park
Y, Wu H. <a href="https://doi.org/10.1093/bioinformatics/btw026" class="external-link">Differential
methylation analysis for <span class="nocase">BS-seq</span> data under
general experimental design</a>. Bioinformatics. 2016
May;32(10):1446–53. </div>
</div>
<div id="ref-korthauerDetectionAccurateFalse2019a" class="csl-entry">
<div class="csl-left-margin">10. </div>
<div class="csl-right-inline">Korthauer K, Chakraborty S, Benjamini Y,
Irizarry RA. <a href="https://doi.org/10.1093/biostatistics/kxy007" class="external-link">Detection and
accurate false discovery rate control of differentially methylated
regions from whole genome bisulfite sequencing</a>. Biostatistics
(Oxford, England). 2019 Jul;20(3):367–83. </div>
</div>
<div id="ref-chenDifferentialMethylationAnalysis2018" class="csl-entry">
<div class="csl-left-margin">11. </div>
<div class="csl-right-inline">Chen Y, Pal B, Visvader JE, Smyth GK. <a href="https://doi.org/10.12688/f1000research.13196.2" class="external-link">Differential
methylation analysis of reduced representation bisulfite sequencing
experiments using <span class="nocase">edgeR</span></a>. F1000Research.
2018 Oct;6:2055. </div>
</div>
<div id="ref-gocukMeasuringInactivationSkew2024" class="csl-entry">
<div class="csl-left-margin">12. </div>
<div class="csl-right-inline">Gocuk SA, Lancaster J, Su S, Jolly JK, Edwards
TL, Hickey DG, et al. <a href="https://doi.org/10.1101/2024.03.20.585856" class="external-link">Measuring
<span>X</span> inactivation skew for retinal diseases with adaptive
nanopore sequencing</a>. bioRxiv; 2024. p. 2024.03.20.585856. </div>
</div>
<div id="ref-royce-tollandArepeatLinksASF2010" class="csl-entry">
<div class="csl-left-margin">13. </div>
<div class="csl-right-inline">Royce-Tolland ME, Andersen AA, Koyfman HR,
Talbot DJ, Wutz A, Tonks ID, et al. <a href="https://doi.org/10.1038/nsmb.1877" class="external-link">The <span class="nocase">A-repeat</span> links <span>ASF</span>/<span class="nocase">SF2-dependent Xist RNA</span> processing with random
choice during <span>X</span> inactivation</a>. Nature Structural &amp;
Molecular Biology. 2010 Aug;17(8):948–54. </div>
</div>
<div id="ref-martinWhatsHapFastAccurate2016" class="csl-entry">
<div class="csl-left-margin">14. </div>
<div class="csl-right-inline">Martin M, Patterson M, Garg S, Fischer SO,
Pisanti N, Klau GW, et al. <a href="https://doi.org/10.1101/085050" class="external-link"><span>WhatsHap</span>: Fast and
accurate read-based phasing</a>. bioRxiv; 2016. p. 085050. </div>
</div>
<div id="ref-deatonCpGIslandsRegulation2011" class="csl-entry">
<div class="csl-left-margin">15. </div>
<div class="csl-right-inline">Deaton AM, Bird A. <a href="https://doi.org/10.1101/gad.2037511" class="external-link"><span>CpG</span> islands and
the regulation of transcription</a>. Genes &amp; Development. 2011
May;25(10):1010–22. </div>
</div>
<div id="ref-ritchieLimmaPowersDifferential2015" class="csl-entry">
<div class="csl-left-margin">16. </div>
<div class="csl-right-inline">Ritchie ME, Phipson B, Wu D, Hu Y, Law CW, Shi
W, et al. <a href="https://doi.org/10.1093/nar/gkv007" class="external-link">Limma powers
differential expression analyses for <span class="nocase">RNA-sequencing</span> and microarray studies</a>. Nucleic
Acids Research. 2015 Apr;43(7):e47. </div>
</div>
<div id="ref-lawrenceSoftwareComputingAnnotating2013" class="csl-entry">
<div class="csl-left-margin">17. </div>
<div class="csl-right-inline">Lawrence M, Huber W, Pagès H, Aboyoun P,
Carlson M, Gentleman R, et al. <a href="https://doi.org/10.1371/journal.pcbi.1003118" class="external-link">Software for
<span>Computing</span> and <span>Annotating Genomic Ranges</span></a>.
PLOS Computational Biology. 2013 Aug;9(8):e1003118. </div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Shian Su.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
